packages/oidc-dpop-client-browser/src/dpop/tokenExchange.spec.ts

```diff
@@ -0,0 +1,430 @@
+ * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+ * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+import URL from "url-parse";
+import { IClient, IIssuerConfig } from "@inrupt/solid-client-authn-core";
+import { it, describe } from "@jest/globals";
+
+import { getTokens, TokenEndpointInput } from "./tokenExchange";
+import { decodeJwt, signJwt } from "./dpop";
+import { JWKECKey } from "jose";
+import { Response } from "cross-fetch";
+
+// Some spec-compliant claims are camel-cased.
+/* eslint-disable @typescript-eslint/camelcase */
+
+// The two following modules introduce randomness in the process, which prevents
+// making assumptions on the returned values. Mocking them out makes keys and
+// DPoP headers predictible.
+jest.mock("./keyGen", () => {
+  return {
+    generateJwkForDpop: (): JWKECKey => {
+      return {
+        kty: "EC",
+        kid: "oOArcXxcwvsaG21jAx_D5CHr4BgVCzCEtlfmNFQtU0s",
+        alg: "ES256",
+        crv: "P-256",
+        x: "0dGe_s-urLhD3mpqYqmSXrqUZApVV5ZNxMJXg7Vp-2A",
+        y: "-oMe9gGkpfIrnJ0aiSUHMdjqYVm5ZrGCeQmRKoIIfj8",
+        d: "yR1bCsR7m4hjFCvWo8Jw3OfNR4aiYDAFbBD9nkudJKM",
+      };
+    },
+  };
+});
+
+// Note that mockJwk and the mocked generateJwkForDpop from keygen return the
+// same mocked key, but due to jest initialisation, one cannot call the other.
+const mockJwk = (): JWKECKey => {
+  return {
+    kty: "EC",
+    kid: "oOArcXxcwvsaG21jAx_D5CHr4BgVCzCEtlfmNFQtU0s",
+    alg: "ES256",
+    crv: "P-256",
+    x: "0dGe_s-urLhD3mpqYqmSXrqUZApVV5ZNxMJXg7Vp-2A",
+    y: "-oMe9gGkpfIrnJ0aiSUHMdjqYVm5ZrGCeQmRKoIIfj8",
+    d: "yR1bCsR7m4hjFCvWo8Jw3OfNR4aiYDAFbBD9nkudJKM",
+  };
+};
+
+jest.mock("uuid", () => {
+  return {
+    v4: (): string => "1234",
+  };
+});
+
+const mockIssuer = (): IIssuerConfig => {
+  return {
+    issuer: new URL("https://some.issuer"),
+    authorizationEndpoint: new URL("https://some.issuer/autorization"),
+    tokenEndpoint: new URL("https://some.issuer/token"),
+    jwksUri: new URL("https://some.issuer/keys"),
+    claimsSupported: ["code", "openid"],
+    subjectTypesSupported: ["public", "pairwise"],
+    registrationEndpoint: new URL("https://some.issuer/registration"),
+    grantTypesSupported: ["authorization_code"],
+  };
+};
+
+const mockWebId = (): string => "https://my.webid";
+
+const mockEndpointInput = (): TokenEndpointInput => {
+  return {
+    grantType: "authorization_code",
+    code: "some code",
+    codeVerifier: "some pkce token",
+    redirectUri: "https://my.app/redirect",
+  };
+};
+
+async function generateMockJwt(): Promise<void> {
+  const payload = {
+    sub: "https://my.webid",
+    iss: mockIssuer().issuer.toString(),
+  };
+  const jwt = await signJwt(payload, mockJwk(), {
+    algorithm: "ES256",
+  });
+  console.log(jwt.toString());
```

Comment:
This might generate a bit too much noise during testing:

```suggestion
  console.log(jwt.toString());
```

Reply:
Fixed in e6751b898d0afafee8dbf01b69e9f9711256ea0b